### ISSUE TITLE
clone() Method Seems to Behave Differently in Derived Classes for Java and Python Bindings

### ISSUE THREAD

**Issue by garrettTuer** [opensim-org/opensim-core#3255]
In Python: The return type matches that of the calling object. For example, when calling the clone method from an object of type Ellipsoid, a "clone" with the type Ellipsoid is returned. I confirmed this by using Python's type() method to check the type of an object returned by this method.

In Java: The return type matches that of the base class of the calling object instance. For example, when calling the clone method from an object of type Ellipsoid, a "clone" with the type OpenSimObject is returned. I confirmed this by checking the return type of the Ellipsoid class's clone() method in the [GUI source code](https://github.com/opensim-org/opensim-gui/blob/master/Gui/opensim/modeling/src/org/opensim/modeling/Ellipsoid.java).

I would think that the intended behavior is that of the Python bindings, with the return type matching the type of the calling object (this seems to match the description of clone() in Object.h). In the case of the Java bindings, I think you'd need to additionally use safeDownCast() to convert the returned value to the desired type, which seems undesirable to me.

**Comment by garrettTuer** [comment]
I haven't found any additional code that affects the behavior of this directive or the produced method in the Java or Python interface files, which makes me suspect the difference is either in SWIG's wrapping tools or in the languages themselves (perhaps something to do with Python's loose typing vs the more strict typing in Java). I'm inclined to think it is the latter and that the reason this has gone "under the radar" is that the main use for the Java code, Matlab, is weakly typed, similar to Python, and does the same potential implicit conversion Python does to return the derived class type when necessary. That's fine for Matlab uses of the Java code, but I think it would provide undesirable behavior in other applications of the Java code, as mentioned at the bottom of the initial post.

**Comment by garrettTuer** [comment]
A few days after I initially posted this issue, I realized that the reason behind this issue is that Java (prior to Java 5) does not support covariant return types. I also noticed that the SWIG interface files for Java bindings ignore the SWIG warning about Java not supporting covariant return types (SWIG warning 822), so I have to assume this issue is already known and isn't considered a high-priority issue due to the main binding applications (Python and MATLAB) being able to implicitly downcast the returned clone to the appropriate type.

It would be nice to be able to generate Java code that uses covariant return types in the clone method for the sake of nicer looking code when using it outside of MATLAB, but, considering the above mentioned acknowledgment of this feature/issue within the SWIG interface files, this might fit more as a feature request rather than an issue. As such, I'll close this issue.
